### DISCUSSION TITLE
Cached page reload is slower than uncached page

### DISCUSSION THREAD

**Discussion by khuezy** [serwist/serwist#163] [General]
Hello, you can see this odd behavior using the example: https://github.com/serwist/serwist/tree/main/examples/next-basic
First edit the `next.config.mjs` file:
```
const withSerwist = withSerwistInit({
  disable: true, // Add me
  ...
});
```

Then run `next dev` and keep reloading the app in the browser, notice there's hardly any delay/latency.

Now remove the `disable` and rerun `next dev`, then reload the page a few times, you'll notice there's a bigger delay/latency.
In the network tab for the `localhost` (home page), the request is in `pending` for 250-500ms or so.

Is this normal? Is there a way to improve this? I remember from the older `next-pwa`, there was `cacheStartUrl`. When I turned that off, it fixed the problem.

```
      - Removed `cacheStartUrl`, `dynamicStartUrl`, and `dynamicStartUrl`. These shall be re-added only when their use cases are made more clear to me.
```
Can this be readded?

Thanks!

**Comment by DuCanhGH** [comment]
@khuezy 

> In the network tab for the localhost (home page), the request is in pending for 250-500ms or so.

That's the service worker processing the request. It is recommended that you disable it during development.

In v10, we may disable Serwist during development by default. Just thinking out loud though.

> Can this be readded?

Nuh uh. They didn't make much sense to me. Also, you fully control your `runtimeCaching` array now, so re-implementing those options isn't really possible either.

**Reply by khuezy** [reply]
I also notice the delay in production.

`Also, you fully control your runtimeCaching array now`
Good point, I'll try this out.

**Reply by khuezy** [reply]
@DuCanhGH `runtimeCaching` doesn't look like it intercepts the root page, it only shows assets such as `js`, `css`, `png/ico`

**Reply by DuCanhGH** [reply]
@khuezy 

> runtimeCaching doesn't look like it intercepts the root page, it only shows assets such as js, css, png/ico

Your service worker intercepts all requests, and Serwist, by default, provides a `runtimeCaching` array that handles any request coming its way. So no, that cannot happen.

> I also notice the delay in production.

That's probably your service worker starting up or processing a request. If that's on its startup time, you should consider using `navigationPreload`. Otherwise, try checking if your matchers, handlers (in Serwist's case, `Strategy` subclasses), or lifecycle callbacks, which usually comes from plugins, take too long to process.

Also, with the service worker intercepting requests, you... pretty much can't expect the speed to match. The service worker may take some time to react, but it saves you the time of going to the network for a response, unless you use `NetworkFirst`/`NetworkOnly`. In that case, not much I can do about the added processing time.

If you have a list of critical pages, consider adding a new `runtimeCaching` entry that strictly matches those pages on top of other entries. This allows said pages to bypass other matchers.

**Reply by khuezy** [reply]
Oh ok, you're right, it does intercept the root `/` page. I had to check `Preserve log` in the dev tool Console, otherwise that log for the root would get cleared before it dumps the rest of the assets (I'm not sure why that happens.)

I have `navigationPreload: true` but that doesn't do much unfortunately. I've tried:
```
runtimeCaching: [
    {
      matcher: ({ url, ...rest }) => {
        return url.pathname === '/'
      },
      handler: new NetworkOnly(),
    },
    ...defaultCache],
 ```
But in the network tab, the root page saying that it's coming from the service worker.

**Reply by khuezy** [reply]
Thanks for the reply and info by the way!
I have a page that streams in additional data after a page loads and this is causing a big delay from the service worker - it blocks the entire page on refresh due to the service worker.

**Reply by DuCanhGH** [reply]
@khuezy 

> But in the network tab, the root page saying that it's coming from the service worker.

The service worker does intercept the request, so the browser reports that the response is from the service worker. Whether it had to go to the network does NOT matter. Hope this helps!

> I have a page that streams in additional data after a page loads and this is causing a big delay from the service worker - it blocks the entire page on refresh due to the service worker.

That sounds problematic... I am not sure what this additional data is though, so I can't be of much help to you. A video or something would be helpful :) If this additional data is loaded after the page is loaded, I don't see how the service worker would block the page... Just the data that gets streamed in.

**Reply by khuezy** [reply]
Sorry! This is actually a nextjs problem. They manage to break streaming on the 15.0.0-canary.11 or higher...

**Reply by DuCanhGH** [reply]
@khuezy heh, things happen... It's a canary build after all.

**Reply by khuezy** [reply]
I just realized why the reload is slower... I had the network inspector tab opened. I don't know why that would cause service workers to load slower. When I close the inspector, the app loads instantly 🤷
